Tests/SGDSolverTests.swift

```diff
@@ -0,0 +1,162 @@
+        let ip2Biases: ValueArray<Float> = [0.60, 0.60]
+
+        let inputs: ValueArray<Float> = [0.05, 0.10]
+        let labels: ValueArray<Float> = [0.01, 0.99]
+
+        let net = Net()
+
+        let data = Source(data: inputs, batchSize: batchSize)
+        let label = Source(data: labels, batchSize: batchSize)
+        let ip1 = InnerProductLayer(weights: ip1Weights, biases: ip1Biases)
+        let sigmoid1 = SigmoidLayer(size: outputSize)
+        let ip2 = InnerProductLayer(weights: ip2Weights, biases: ip2Biases)
+        let sigmoid2 = SigmoidLayer(size: outputSize)
+        let loss = L2LossLayer(size: labelSize)
+        let sink = Sink(inputSize: 1, batchSize: batchSize)
+
+        let dataLayer = net.addLayer(data, name: "data")
+        let labelLayer = net.addLayer(label, name: "label")
+        let ip1Layer = net.addLayer(ip1, name: "ip 1")
+        let sigmoid1Layer = net.addLayer(sigmoid1, name: "sigmoid 1")
+        let ip2Layer = net.addLayer(ip2, name: "ip 2")
+        let sigmoid2Layer = net.addLayer(sigmoid2, name: "sigmoid 2")
+        let lossLayer = net.addLayer(loss, name: "loss")
+        let sinkLayer = net.addLayer(sink, name: "sink")
+
+        let ip1Buffer = net.addBufferWithName("ip 1")
+        let sigmoid1Buffer = net.addBufferWithName("sigmoid 1")
+        let ip2Buffer = net.addBufferWithName("ip 2")
+        let sigmoid2Buffer = net.addBufferWithName("sigmoid 2")
+        let lossBuffer = net.addBufferWithName("loss")
+        let sinkBuffer = net.addBufferWithName("sink")
+
+        net.connectLayer(dataLayer, toBuffer: ip1Buffer)
+        net.connectBuffer(ip1Buffer, toLayer: ip1Layer)
+        net.connectLayer(ip1Layer, toBuffer: sigmoid1Buffer)
+        net.connectBuffer(sigmoid1Buffer, toLayer: sigmoid1Layer)
+        net.connectLayer(sigmoid1Layer, toBuffer: ip2Buffer)
+        net.connectBuffer(ip2Buffer, toLayer: ip2Layer)
+        net.connectLayer(ip2Layer, toBuffer: sigmoid2Buffer)
+        net.connectBuffer(sigmoid2Buffer, toLayer: sigmoid2Layer)
+        net.connectLayer(sigmoid2Layer, toBuffer: lossBuffer, atOffset: 0)
+        net.connectLayer(labelLayer, toBuffer: lossBuffer, atOffset: sigmoid2.outputSize)
+        net.connectBuffer(lossBuffer, toLayer: lossLayer)
+        net.connectLayer(lossLayer, toBuffer: sinkBuffer)
+        net.connectBuffer(sinkBuffer, toLayer: sinkLayer)
+
+        let sigmoid2BufferId = net.nodes.reduce(-1) { val, node in
+            if node.layer is SigmoidLayer {
+                return node.outputBuffer!.id
+            }
+            return val
+        }
+
+        let solver = try! SGDSolver(device: device, net: net, batchSize: batchSize, endStep: 1, learningRate: 0.5, momentum: 1.0)
+        let expecation1 = self.expectationWithDescription("Net forward pass 1")
+        let expecation2 = self.expectationWithDescription("Net forward pass 2")
+        solver.train({
+            solver.completion = { snapshot in
```

Comment:
@aleph7 because the `completion` in solver will wipe buffers before I have a chance to access. That's why I'm overwriting it with my assertions.